crates/turbo-tasks-fs/src/json.rs

```diff
@@ -0,0 +1,120 @@
+use std::{
+    borrow::Cow,
+    fmt::{Display, Formatter, Write},
+};
+
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+use turbo_tasks::trace::TraceRawVcs;
+
+use crate::{rope::Rope, source_context::get_source_context};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, TraceRawVcs)]
+pub struct UnparseableJson {
+    #[turbo_tasks(trace_ignore)]
+    pub message: Cow<'static, str>,
+    pub path: Option<String>,
+    /// The start line and column of the error.
+    /// Line and column is 0-based.
+    pub start_location: Option<(usize, usize)>,
+    /// The end line and column of the error.
+    /// Line and column is 0-based.
+    pub end_location: Option<(usize, usize)>,
+}
+
+/// Converts a byte position to a 0-based line and column.
+fn byte_to_location(pos: usize, text: &str) -> (usize, usize) {
+    let text = &text[..pos];
+    let mut lines = text.lines().rev();
+    let last = lines.next().unwrap_or("");
+    let column = last.len();
+    let line = lines.count();
+    (line, column)
+}
+
+impl UnparseableJson {
+    pub fn from_jsonc_error(e: jsonc_parser::errors::ParseError, text: &str) -> Self {
+        Self {
+            message: e.message.clone().into(),
+            path: None,
+            start_location: Some(byte_to_location(e.range.start, text)),
+            end_location: Some(byte_to_location(e.range.end, text)),
+        }
+    }
+
+    pub fn from_serde_path_to_error(e: serde_path_to_error::Error<serde_json::Error>) -> Self {
+        let inner = e.inner();
+        Self {
+            message: inner.to_string().into(),
+            path: Some(e.path().to_string()),
+            start_location: Some((inner.line() - 1, inner.column() - 1)),
+            end_location: None,
+        }
+    }
+
+    pub fn write_with_content(&self, writer: &mut impl Write, text: &str) -> std::fmt::Result {
+        writeln!(writer, "{}", self.message)?;
+        if let Some(path) = &self.path {
+            writeln!(writer, "  at {}", path)?;
```

Comment:
I think the `at` is redundant here. We can remove it and the error would read more easily.